agent/consul/stream/topic_key.go

```diff
@@ -0,0 +1,36 @@
+package stream
+
+import "strings"
+
+// TopicKey identifies a portion of a topic for which a subscriber wishes to
+// receive events (e.g. health events for a particular service). It is used as
+// a map key when storing/retrieving event buffers and cached snapshots.
+//
+// Note: its fields are unexported as it's expected that users will call
+// NewTopicKey rather than constructing a TopicKey manually, as this method
+// normalizes the given values.
+type TopicKey struct {
+	key       string
+	namespace string
+	partition string
+}
+
+// NewTopicKey constructs a TopicKey for the given {key, namespace, partition}
+// combination. The given values will be normalized - converted to lower-case
+// with empty namespace/partition treated as the default namespace/partition -
+// to ensure equality when using the TopicKey as a map key.
+func NewTopicKey(key, namespace, partition string) TopicKey {
+	if namespace == "" {
+		namespace = "default"
```

Comment:
I forget where we landed on hard coding the string `default` vs it being a constant somewhere. I know at one point it was only a constant in some awkward places and I don't think it's worth taking a dependency on some huge irrelevant package for that, but perhaps we fixed that and have a clean import we could use rather than hard coding this in another place. @dnephin probably knows if there is a better alternative!

Reply:
I don't think the `stream` package should know anything about the default namespace. This defaulting should have already happened long before we get here, because it must happen either when we accept a subscription or before we publish an event.  I think `TopicKey` should be an opaque identifier.